[BACKEND] Adding optional environment variable for dumping reproducer mlir files

plotfi opened the pull request:
This PR is a work in progress and depends on https://github.com/llvm/llvm-project/pull/75421 so only can be landed after Triton bumps it's llvm hash.

This PR makes it possible to dump an MLIR reproducer when setting the env var TRITON_REPRODUCER_PATH to some mlir file path (`export TRITON_REPRODUCER_PATH=$(PWD)/triton-reproducer.mlir`). The reproducer can then be run using triton-opt like so: `triton-opt  --run-reproducer ./triton-reproducer.mlir`

TritonLLVMIR has been added as a link dependency for the bin triton tools because that is needed for properly registering the enable-line-info pass (which gets inserted into the pass pipeline if `export TRITON_DISABLE_LINE_INFO=1` is not set). The enable-line-info pass is also added for registration in RegisterTritonDialects.h as well. 

@ThomasRaoux

ThomasRaoux commented on bin/RegisterTritonDialects.h:
```diff
@@ -36,6 +38,9 @@ inline void registerTritonDialects(mlir::DialectRegistry &registry) {
   mlir::triton::registerConvertTritonToTritonGPUPass();
   mlir::triton::registerConvertTritonGPUToLLVMPass();
   mlir::triton::registerConvertNVGPUToLLVMPass();
+  if (!::triton::tools::getBoolEnv("TRITON_DISABLE_LINE_INFO")) {
+    mlir::registerLLVMDIScope();
+  }
```
how is this related? Is there any reason to not always call `mlir::registerLLVMDIScope();`?

plotfi replied:
Yeah, I think this was silly on my part. It should register the pass always in the bin tools I think.

ThomasRaoux commented on python/src/ir.cc:
```diff
@@ -1545,6 +1545,15 @@ void init_triton_ir(py::module &&m) {
       .def("run", [](mlir::PassManager &self, mlir::ModuleOp &mod) {
         // TODO: maybe dump module to file and print error for better
         // diagnostics
+        if (auto reproducerPath =
+            ::triton::tools::getenv("TRITON_REPRODUCER_PATH");
+            !reproducerPath.empty()) {
```
nit: the syntax is a bit weird, I would move `auto reproducerPath =           ::triton::tools::getenv("TRITON_REPRODUCER_PATH");` above the `if`

plotfi commented on python/src/ir.cc:
```diff
@@ -1545,6 +1545,15 @@ void init_triton_ir(py::module &&m) {
+          auto anchorName = self.getOpAnchorName();
+          auto passes = self.getPasses();
+          mlir::Operation* op = mod.getOperation();
+          mlir::makeReproducer(anchorName, passes, op, reproducerPath);
```
Can we do this in a future PR?

htyu replied:
Sounds good!

ThomasRaoux commented on python/test/unit/language/test_reproducer.py:
```diff
@@ -0,0 +1,40 @@
+    return
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="requires cuda")
+def test_line_info():
```
this doesn't seem like the right name?

plotfi replied:
ah, yeah copy-paste error.

ThomasRaoux commented on python/test/unit/language/test_reproducer.py:
```diff
@@ -0,0 +1,40 @@
+        raise Exception("Failed to find pipeline info in reproducer file.")
+    if not "convert-triton-gpu-to-llvm" in foundPipeline:
+        raise Exception("Failed to find triton passes in pipeline")
+    print(foundPipeline)
```
is this left over debug?

ThomasRaoux commented:
LLVM got integrated, can you run pre-commit on this?

plotfi commented:
> LLVM got integrated, can you run pre-commit on this?

Updated branch, ran `pre-commit run --all-files` and updated based on what it changed. Everything looks green now.
